database/neo4j/neo4j.go

```diff
@@ -0,0 +1,232 @@
+package neo4j
+
+import (
+	"C" // import C so that we can't compile with CGO_ENABLED=0
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	neturl "net/url"
+	"sync/atomic"
+
+	"github.com/golang-migrate/migrate/v4/database"
+	"github.com/neo4j/neo4j-go-driver/neo4j"
+)
+
+func init() {
+	db := Neo4j{}
+	database.Register("neo4j", &db)
+}
+
+var DefaultMigrationsLabel = "SchemaMigration"
+
+var (
+	ErrNilConfig = fmt.Errorf("no config")
+)
+
+type Config struct {
+	AuthToken       neo4j.AuthToken
+	URL             string // if using WithInstance, don't provide auth in the URL, it will be ignored
+	MigrationsLabel string
+}
+
+type Neo4j struct {
+	driver neo4j.Driver
+	lock   uint32
+
+	// Open and WithInstance need to guarantee that config is never nil
+	config *Config
+}
+
+func WithInstance(config *Config) (database.Driver, error) {
+	if config == nil {
+		return nil, ErrNilConfig
+	}
+
+	neoDriver, err := neo4j.NewDriver(config.URL, config.AuthToken)
+	if err != nil {
+		return nil, err
+	}
+
+	driver := &Neo4j{
+		driver: neoDriver,
+		config: config,
+	}
+
+	if err := driver.ensureVersionConstraint(); err != nil {
+		return nil, err
+	}
+
+	return driver, nil
+}
+
+func (n *Neo4j) Open(url string) (database.Driver, error) {
+	uri, err := neturl.Parse(url)
+	if err != nil {
+		return nil, err
+	}
+	password, _ := uri.User.Password()
+	authToken := neo4j.BasicAuth(uri.User.Username(), password, "")
+	uri.User = nil
+	uri.Scheme = "bolt"
+
+	return WithInstance(&Config{
+		URL:             uri.String(),
+		AuthToken:       authToken,
+		MigrationsLabel: DefaultMigrationsLabel,
+	})
+}
+
+func (n *Neo4j) Close() error {
+	return n.driver.Close()
+}
+
+// local locking in order to pass tests, Neo doesn't support database locking
+func (n *Neo4j) Lock() error {
+	if !atomic.CompareAndSwapUint32(&n.lock, 0, 1) {
+		return database.ErrLocked
+	}
+
+	return nil
+}
+
+func (n *Neo4j) Unlock() error {
+	if !atomic.CompareAndSwapUint32(&n.lock, 1, 0) {
+		return database.ErrNotLocked
+	}
+	return nil
+}
+
+func (n *Neo4j) Run(migration io.Reader) error {
+	body, err := ioutil.ReadAll(migration)
+	if err != nil {
+		return err
+	}
+
+	session, err := n.driver.Session(neo4j.AccessModeWrite)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err := session.Close()
+		if err != nil {
+			log.Printf("error: %s", err)
```

Comment:
Use named return values to handle errors in deferred method calls.
https://blog.learngoprogramming.com/5-gotchas-of-defer-in-go-golang-part-iii-36a1ab3d6ef1#8869

Reply:
Done